### ISSUE TITLE
Critical EMA implementation bug, resulting in dead training.

### ISSUE THREAD

**Issue by Adamusen** [WongKinYiu/YOLO#97]
## Describe the bug
The Exponential Moving Average (EMA) implementation is missing a critical component in the yolo/tools/solver.py/ModelTrainer class, in which its update() function is never called. This leads to the fact, that the weights of the model are only updated in it's "shadow" copy of each model parameter once (see yolo/utils/model_utils.py/ExponentialMovingAverage class), at the beginning of the training process.

During checkpointing, however, if EMA is enabled (which is by default, see yolo/config/task/train.yaml), these only at the beginning stored weights (either pretrained weights or random weights if starting from scratch) are copied back and overwriting the now for one epoch trained model's parameters (see Lines 120-123 in yolo/tools/solver.py), right before the checkpoint is created.

Because of another problem with ema.apply_shadow() and ema.restore() (described later), this leads to the fact, that both the "model_state_dict" and "model_state_dict_ema" weights saved in the checkpoint are just the same, with which the training process was started (every time in every ckpt), with all the progress achieved by the training itself being lost. It is not only lost in the saved checkpoint, but in general, which means, after each epoch, which creates a checkpoint, the model parameters reset. I'm not sure, but this might explain the "slow convergence speed" issue currently at the top of the README.md

## Additional related problems
The following code (Lines 115-126 in yolo/tools/solver.py) does not work as intended.
```
        checkpoint = {
            "epoch": epoch_idx,
            "model_state_dict": self.model.state_dict(),
            "optimizer_state_dict": self.optimizer.state_dict(),
        }
        if self.ema:
            self.ema.apply_shadow()
            checkpoint["model_state_dict_ema"] = self.model.state_dict()
            self.ema.restore()

        logger.info(f"💾 success save at {file_path}")
        torch.save(checkpoint, file_path)
```
Since the created dictionary (checkpoint) is just a collection of references to sets of Tensors, that are only saved on disk when calling torch.save(), at the time of saving both "model_state_dict" and "model_state_dict_ema" will have the same values, as they are both just references to the same set of tensors contained in self.model.state_dict(). To truly save both set of tensor (the state_dict and ema.shadow), deep copies of the tensors must be created and referenced in the checkpoint dictionary before saving.

This leads to another problem, namely, that the ema.apply_shadow() function does not save the original state_dict before copies the ema.shadow values in to it (see yolo/utils/model_utils.py/ExponentialMovingAverage class implementation), for this reason the call to ema.restore() does nothing being called after ema.apply_shadow(). This is an additional main cause of the original bug, in which all the progress of training is eventually lost when executing the code snippet above.

## To Reproduce
1. Train a network with EMA enabled.
2. Open your saved checkpoints with Netron or a similar NN visualizing app.
3. Observe the weights being the same initial ones in every ckpt.

## Expected behavior
1. The "shadow" copy of the model parameters in the EMA to be updated during training after each step.
2. The ema.apply_shadow() and ema.restore() functions to properly restore the original model.state_dict()
3. The values of model_state_dict and model_state_dict_ema to be saved in the checkpoint correctly (not the same values twice)

**Comment by Adamusen** [comment]
Found the commit, in which the missing ema.update() got removed:
[https://github.com/WongKinYiu/YOLO/commit/b4bcccbc8a3180385577f12613175c45df08fd70](url)

**Comment by DmitriiSavin** [comment]
I started training on a custom dataset and counted the loss on the train after each epoch

Validation and saving of the checkpoint occurred every 10 epochs.

As you can see, the model rolled back after each validation


![image](https://github.com/user-attachments/assets/3b1e9ff3-aab9-4043-8ca3-bb723927aa14)

**Comment by Adamusen** [comment]
For anyone trying to train a network at the current state, as a temporary workaround, turn EMA off for now until the issue is fixed by setting it to False in yolo/config/task/train.yaml. Training and checkpointing works properly as long as it is disabled :)

```
ema:
  enabled: False
  decay: 0.995
```

**Comment by Adamusen** [comment]
Implemented a fix for the issue and opened a pull request for the changes.
(Sorry for the funny looking commit history, first time doing this :) )

**Comment by nkinnaird** [comment]
It looks like this was all replaced when the lightning framework was implemented for training:

https://github.com/WongKinYiu/YOLO/commit/8b3b3ef01d9aa8388f6d45ce3138e2d4daadee2a#diff-d81fac5caf45e1f790eb45f09f723cb5ea25adfcb3c55f5988075dfe34523da4L69

Not sure what that means in terms of the EMA functionality in general.
